keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+                storage: 250Gi
+ ```
+ 
+Form an infrastructure management perspective, the manifest above encapsulates everything that is necessary to create 
+the workloads and services that compose our WordPress CMS application. However, the fact that these workloads comprise 
+an application is lost in translation during creation. Once these manifests are applied, there is no default way 
+for users to reconstruct the fact that the whole is more than the sum of its parts.
+ 
+When UIs, either command line or graphical, seek to interact with the application, they can only interact with its 
+components, and there is no standard way to indicate that those components comprise an application. As previously 
+mentioned, Many tools in the ecosystem attempt to aggregate the components of an application using labels and/or 
+annotations, but, as each tool uses a different scheme, an application created by one tool is generally not recognizable 
+by another, and UI designers have no standard way to recognize and display an application created by any tool.
+
+Moreover, their is no way to determine if the application is healthy. Even if a UI is able to determine the components 
+and dependencies of the application, and even if it is able to determine that the individual components are healthy 
+(e.g All Services are created and all Deployments are fully replicated), this does not necessarily imply that the 
+application as a whole is able to service requests.
+
+
+### Goals
+
+1. Provide a standard API for creating, viewing, and managing applications in Kubernetes.
+1. Provide a CLI implementation, via kubectl, that interacts with the Application API.
+1. Provide installation status and garbage collection for applications.
+1. Provide a standard way for applications to surface a basic health check to the UIs.
+1. Provide an explicit mechanism for applications to declare dependencies on another application.
+1. Promote interoperability among ecosystem tools and UIs by creating a standard that tools MAY implement.
+
+### Non-Goals
+
+1. Create a standard that all tools MUST implement. 
+1. Provide a way for UIs to surface metrics from an application.
+
+## Proposal
+In order to address the issues discussed in the [previous section](#motivation) we will introduce the Application Kind 
+to the apps Group of the Kubernetes API, add the Application Controller to the Controller Manager, and modify 
+kubectl to make use the of the Application Kind.
+
+### User Stories
+
+#### Application Creation
+As an application maintainer or tool developer, I can create an application that is interoperable with other tools and 
+has first class support in Kuberentes UIs.
+
+#### Application View
+As an application user or UI designer, I have a standard way to view the components and status of Kubernetes 
+applications.
+
+#### Application Health
+As an application developer, I can surface a basic health check for UIs to consume, and, as a UI developer, I have a 
```

Comment:
why do UI's want to consume basic health check @kow3ns ?

Reply:
@krmayankk A UI could display if the app is healthy or not with a single request rather than having to query all objects in the app. This is fewer HTTP(S) requests outside of the cluster in many cases.

Reply:
Yes, an application needs to provide the mechanism (or we need to define a mehcanism) to be able to communicate a high-level (application) health state.

Think of weave or alike, which has multiple pods - i.e. DaemonSet like handlers - you want one place to check it's health.